homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,236 @@
+
+    if connection_type == CONF_CONNECTION_TYPE_NIBEGW:
+        connection = NibeGW(
+            heatpump,
+            entry.data[CONF_IP_ADDRESS],
+            entry.data[CONF_REMOTE_READ_PORT],
+            entry.data[CONF_REMOTE_WRITE_PORT],
+            listening_port=entry.data[CONF_LISTENING_PORT],
+        )
+    else:
+        raise HomeAssistantError(f"Connection type {connection_type} is not supported.")
+
+    await connection.start()
+    try:
+        coordinator = Coordinator(hass, heatpump, connection)
+
+        data = hass.data.setdefault(DOMAIN, {})
+        data[entry.entry_id] = coordinator
+        await coordinator.async_config_entry_first_refresh()
+
+        reg = dr.async_get(hass)
+        reg.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            identifiers={(DOMAIN, entry.unique_id or entry.entry_id)},
+            manufacturer="NIBE Energy Systems",
+            model=heatpump.model.name,
+            name=heatpump.model.name,
+        )
+
+        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    except Exception:
+        await connection.stop()
+        raise
+
+    # Trigger a refresh again now that all platforms have registered
+    hass.async_create_task(coordinator.async_refresh())
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        coordinator: Coordinator = hass.data[DOMAIN].pop(entry.entry_id)
+        await coordinator.connection.stop()
+
+        # sleep a bit to make sure sockets are closed
+        await asyncio.sleep(5)
+
+    return unload_ok
+
+
+class Coordinator(DataUpdateCoordinator[dict[int, Coil]]):
+    """Update coordinator for nibe heat pumps."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        heatpump: HeatPump,
+        connection: Connection,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, LOGGER, name="Nibe Heat Pump", update_interval=timedelta(seconds=60)
+        )
+
+        self.connection = connection
+        self.heatpump = heatpump
+
+    @property
+    def coils(self) -> list[Coil]:
+        """Return the full coil database."""
+        return self.heatpump.get_coils()
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this coordinator."""
+        return self.config_entry.unique_id or self.config_entry.entry_id
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for the main device."""
+        return DeviceInfo(identifiers={(DOMAIN, self.unique_id)})
+
+    def get_coil_data(self, coil: Coil | None) -> Coil | None:
+        """Return a coil with data and check for validity."""
+        if not self.data or not coil:
```

Comment:
Can we just define this `data` in `__init__()`?

Reply:
yup